File: composeApp/src/iosMain/kotlin/a/atiyah/sa/exploreai/ask_ai/domain/GenerativeModelIOS.kt
package a.atiyah.sa.exploreai.ask_ai.domain

import a.atiyah.sa.exploreai.di.generativeModelIOS

class GenerativeModelIOS: GenerativeModel {
    override suspend fun generateTextContent(prompt: String): String? {
        return generativeModelIOS?.generateTextContent(prompt)
    }

    override suspend fun generateJsonContent(prompt: String): String? {
        return generativeModelIOS?.generateJsonContent(prompt)
    }
}

File: iosApp/iosApp/ask_ai/domain/GenerativeModelIOS.swift
//
//  GenerativeModelIOS.swift
//  iosApp
//
//  Created by Abdulrahman on 15/06/1446 AH.
//  Copyright © 1446 AH orgName. All rights reserved.
//
import ComposeApp
import FirebaseCore
import FirebaseVertexAI

class GenerativeModelIOS: ComposeApp.GenerativeModel {
    static let shared = GenerativeModelIOS()

    let vertex = VertexAI.vertexAI()

    let jsonSchema = Schema.array(
      items: .object(
        properties: [
          "name": .string(),
          "country": .string()
        ]
      )
    )

    func generateTextContent(prompt: String) async throws -> String? {
        let model = vertex.generativeModel(
            modelName: "gemini-1.5-flash"
        )

        return try await model.generateContent(prompt).text
    }


    func generateJsonContent(prompt: String) async throws -> String? {
        let model = vertex.generativeModel(
            modelName: "gemini-1.5-flash",
            generationConfig: GenerationConfig(
                responseMIMEType: "application/json",
                responseSchema: jsonSchema
            )
        )

        return try await model.generateContent(prompt).text
    }
}
